Dockerfile

```
@@ -1,44 +1,54 @@
-FROM python:3.7.3-alpine
+FROM python:3.7-slim AS python-builder
 
-# Default to building for local development.
-# Override with `--build-arg PIP_FILE=build.txt` to build for production.
-ARG PIP_FILE=all.txt
-
-EXPOSE 8000
+ENV PIP_DISABLE_PIP_VERSION_CHECK=1
+ENV PATH="/venv/bin:$PATH"
 
 WORKDIR /app
 
-RUN addgroup -g 10001 app && \
-    adduser -D -u 10001 -G app -h /app -s /sbin/nologin app
-
-RUN apk --no-cache add \
-    build-base \
-    bash \
-    curl \
-    git \
+RUN apt-get update && \
+    apt-get -y upgrade && \
+    apt-get -y install \ 
+    libpq-dev \
     libffi-dev \
-    py-cffi \
-    postgresql \
-    postgresql-dev \
-    postgresql-client
-
-# Install Python dependencies
-COPY requirements/*.txt /tmp/requirements/
-# Switch to /tmp to install dependencies outside home dir
-WORKDIR /tmp
-# TODO: Consider a way to install only the "build.txt" deps for production.
-RUN pip install --require-hashes --no-cache-dir -r requirements/$PIP_FILE
+    python-dev \
+    build-essential
+
+# Install virtualenv
+RUN pip install virtualenv
+RUN virtualenv /venv
+
+# Install base python dependencies 
+COPY requirements/*.txt ./requirements/
+RUN pip install --require-hashes --no-cache-dir -r requirements/build.txt
+
+FROM python:3.7-slim AS base
+
+ENV PIP_DISABLE_PIP_VERSION_CHECK=1
+ENV PATH="/venv/bin:$PATH"
+
+RUN groupadd --gid 10001 app && useradd -g app --uid 10001 --shell /usr/sbin/nologin app
+
+COPY --from=python-builder /venv /venv
+
+EXPOSE 8000
 
 WORKDIR /app
 COPY . /app
-RUN chown app:app -R .
-USER app
 
+# Collect static files
 RUN DEBUG=False SECRET_KEY=foo ./manage.py collectstatic --noinput -c
 
+# Set user permissions
+COPY --chown=app:app . . 
+USER app
 # Using /bin/bash as the entrypoint works around some volume mount issues on Windows
 # where volume-mounted files do not have execute bits set.
 # https://github.com/docker/compose/issues/2301#issuecomment-154450785 has additional background.
 ENTRYPOINT ["/bin/bash", "/app/bin/run"]
 
-CMD ["dev"]
+FROM base AS devapp
+
+USER root
+# Install dev python dependencies
+RUN pip install --require-hashes --no-cache-dir -r requirements/tests.txt
```

We should install everything now: some of the test dependencies now depend on some main dependencies, which means their hashes need to be specified. No harm in this, as pip is smart enough not to reinstall anything already installed.

```suggestion
RUN pip install --require-hashes --no-cache-dir -r requirements/all.txt
```